mrmustard/math/lattice/strategies/beamsplitter.py

```diff
@@ -0,0 +1,156 @@
+# Copyright 2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+# Recurrencies for Fock-Bargmann amplitudes put together a strategy for
+# enumerating the indices in a specific order and functions for calculating
+# which neighbours to use in the calculation of the amplitude at a given index.
+# In summary, they return the value of the amplitude at the given index by following
+# a recipe made of two parts. The function to recompute A and b is determined by
+# which neighbours are used.
+
+"""Fock-Bargmann recurrence relation steps optimized for beamsplitter."""
+
+import numpy as np
+from numba import njit
+
+from mrmustard.math.lattice import steps
+from mrmustard.typing import ComplexMatrix, ComplexTensor, ComplexVector
+
+SQRT = np.sqrt(np.arange(100000))
+
+__all__ = ["beamsplitter", "beamsplitter_vjp"]
+
+
+@njit
+def beamsplitter(
+    shape: tuple[int, int, int, int], theta: float, phi: float, dtype=np.complex128
+) -> ComplexTensor:  # pragma: no cover
+    r"""Calculates the Fock representation of the beamsplitter.
+    It takes advantage of input-output particle conservation (m+n=p+q)
+    to avoid one for loop. Inspired from the original implementation in
+    the walrus by @ziofil. Here is how the parameters are used in the
+    code:
```

Comment:
Maybe we can point out this is in your paper Equation (73) to (75)?

Reply:
good idea

Reply:
```suggestion
    code (see eq. 73-75 in https://arxiv.org/abs/2004.11002):
```